File: solutions/c/148.c
static struct ListNode *
mergeList(struct ListNode *l1, struct ListNode *l2);

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *
sortList(struct ListNode *head)
{
    if (head == NULL || head->next == NULL) {
        return head;
    }

    struct ListNode *slow, *fast, *tmp;
    tmp = slow = fast = head;

    while (fast != NULL && fast->next != NULL) {
        tmp = slow;
        slow = slow->next;
        fast = fast->next->next;
    }

    tmp->next = NULL;

    struct ListNode *l1 = sortList(head);
    struct ListNode *l2 = sortList(slow);

    return mergeList(l1, l2);
}

struct ListNode *
mergeList(struct ListNode *l1, struct ListNode *l2)
{
    struct ListNode dummy;
    struct ListNode *tail = &dummy;

    while (l1 != NULL || l2 != NULL) {
        if (l1 == NULL) {
            tail->next = l2;
            break;
        } else if (l2 == NULL) {
            tail->next = l1;
            break;
        } else {
            if (l1->val < l2->val) {
                tail->next = l1;
                l1 = l1->next;
            } else {
                tail->next = l2;
                l2 = l2->next;
            }

            tail = tail->next;
        }
    }

    return dummy.next;
}

File: solutions/go/148.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func sortList(head *ListNode) *ListNode {
	return qsort(head, nil)
}

func qsort(head, tail *ListNode) *ListNode {
	if head == nil || head.Next == nil || head == tail {
		return head
	}

	pivot := partition(head, tail)
	qsort(head, pivot)
	qsort(pivot.Next, tail)

	return head
}

func partition(head, tail *ListNode) *ListNode {
	pivot := head
	tmp := head.Next

	for tmp != nil && tmp != tail {
		if tmp.Val < head.Val {
			pivot = pivot.Next
			tmp.Val, pivot.Val = pivot.Val, tmp.Val
		}
		tmp = tmp.Next
	}

	head.Val, pivot.Val = pivot.Val, head.Val
	return pivot
}

// sortListV2 Another version using merge sort
func sortListV2(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	var tmp *ListNode
	slow, fast := head, head

	for fast != nil && fast.Next != nil {
		tmp = slow
		slow = slow.Next
		fast = fast.Next.Next
	}

	tmp.Next = nil

	l1 := sortListV2(head)
	l2 := sortListV2(slow)

	return mergeList(l1, l2)
}

func mergeList(l1, l2 *ListNode) *ListNode {
	var dummy ListNode
	ptr := &dummy

	for l1 != nil || l2 != nil {
		if l1 != nil && l2 != nil {
			if l1.Val < l2.Val {
				ptr.Next = l1
				l1 = l1.Next
			} else {
				ptr.Next = l2
				l2 = l2.Next
			}
			ptr = ptr.Next
		} else if l1 != nil {
			ptr.Next = l1
			break
		} else {
			ptr.Next = l2
			break
		}
	}

	return dummy.Next
}
